### PR TITLE
Recurrence Exceptions

### PR DESCRIPTION
**Pull request by ChrisJAllan**
For some reason the server would reject exceptions that had a body or an empty category list, so this is unable to change the description or erase all categories. Everything else seems to work: changing the title, changing the time, adding or changing categories, deleting an occurrence, etc.

### CODE REVIEW

**Comment by jobisoft** [comment]
Thanks a lot for your work!
What version of active sync did you test this with, 2.5 or 14.0?

The problem of body and categories *might* be due to wrong order inside the wbxml. Could you check, at what position the server places categories and body tags and adjust the wbxml you are creating?

**Comment by jobisoft** [comment]
Also it would be interesting to see, how your server is sending "clear all categories", when you clear all categories on the server and sync that to Thunderbird.

Same goes for the cleared body.

File: content/provider/eas/calendarsync.js [hunk 1]
```diff
@@ -316,12 +340,17 @@ eas.sync.Calendar = {
         
         //each TB event has an ID, which is used as EAS serverId - however there is a second UID in the ApplicationData
         //since we do not have two different IDs to use, we use the same ID
-        wbxml.atag("UID", item.id);
+        if (!(isException && asversion != "2.5")) {
```
Comment (jobisoft) [inline content/provider/eas/calendarsync.js]: Check if UID is required in 2.5 or optional, do not specify at all?

File: content/provider/eas/calendarsync.js [hunk 2]
```diff
@@ -304,6 +325,9 @@ eas.sync.Calendar = {
     //read TB event and return its data as WBXML
     // --------------------------------------------------------------------------- //
     getWbxmlFromThunderbirdItem: function (item, syncdata) {
```
Comment (jobisoft) [inline content/provider/eas/calendarsync.js]: do not introduce new function, but use default value for isException = false

File: content/provider/eas/calendarsync.js [hunk 3]
```diff
@@ -340,22 +369,26 @@ eas.sync.Calendar = {
...
+            if (item.organizer && item.organizer.id) wbxml.atag("OrganizerEmail",  cal.removeMailTo(item.organizer.id));
+        }
 
         //Attendees
         let TB_responseType = null;
         let countAttendees = {};
         let attendees = item.getAttendees(countAttendees);
         
-        if (countAttendees.value > 0) {
+        if (countAttendees.value > 0 && !(isException && asversion == "2.5")) {
```
Comment (jobisoft) [inline content/provider/eas/calendarsync.js]: for better readability, move if !(isException && asversion == "2.5") as outer if.

**Comment by ChrisJAllan** [comment]
I am using 14.0. I tried using 2.5, but it wouldn't sync anything. I'm using an office365 outlook account though my school, so that may have something to do with it. I'll look into testing with z-push today.

[Here's a log of the server sending an exception with a description and empty category list.](https://github.com/jobisoft/TbSync/files/1649189/debug.copy.log)

[Sending the same event back and failing.](https://github.com/jobisoft/TbSync/files/1649195/debug.another.copy.log)

**Comment by jobisoft** [comment]
sorry for moving everything around again ...

**Comment by jobisoft** [comment]
please re-check the body problem with latest code, I removed NativeBodyType from Body which caused issues with tasks as well. the if (!isException) is still in though.

**Comment by ChrisJAllan** [comment]
That works! Looking through the docs, `NativeBodyType` is not allowed in exceptions. It's also optional for standard elements and seems to do the same thing as Body->Type, so I don't see any reason to keep it.

**Comment by jobisoft** [comment]
Done. Open question:

UID in Exception for 2.5 is unique, or is it the same UID as the item containing this exception?

Organizer is not allowed in exception I think. But does the Thunderbird Exception item (I think you called that replacement) needs an Organizer? Should we use the same as in the "parent" item (containing this exception)? If so, we must not pass "isException" but "parentItem" (defaults to null) and if that is present, we know it is an exception and also have access to the parent item, what do you think?

createWBXML has now an optional parameter to set the initial codepage (up to now it always started with AirSync). If we use wbxml.append (which is a nice idea, thanks!) we need to tell him at what codepage we are currently at. Example:
https://github.com/jobisoft/TbSync/blob/master/content/provider/eas/sync.js#L66

because this call:
https://github.com/jobisoft/TbSync/blob/master/content/provider/eas/calendarsync.js#L202

is already in Calendar codepage:
https://github.com/jobisoft/TbSync/blob/master/content/provider/eas/calendarsync.js#L174

What about this call:
https://github.com/jobisoft/TbSync/blob/master/content/provider/eas/sync.js#L328

At current getWbxmlFromThunderbirdItem thinks it is in AirSync and switches to Calendar/Tasks and at the end back to AirSync. Do we actually need those switches? What codepage are we actually at at sync.js#L328?

**Comment by ChrisJAllan** [comment]
1. No idea. The docs aren't particularly helpful and I haven't found a way to test 2.5 yet.
2. The replacement already inherits all the values from the parent because it comes from `item.recurrenceInfo.getOccurrenceFor(dateTime)`. `setThunderbirdItemFromWbxml` then only modifies things that are different then the parent.
3. It might be useful to have `wbxml` ignore switches to the current page. Then, instead of using `append`, pass `wbxml` as an argument. Also, it doesn't seem that Tasks support exceptions, so L328 does work, though it might be better to change it to `syncdata.type` anyway.

**Comment by jobisoft** [comment]
Thanks again for your great work. I have not yet fully understood your code, but will look into it after I finished task support.

However, I just realized that EAS v2.5 support was broken, it is now fixed and I get calendars and tasks with 2.5. If you do not want to spend any more time on 2.5, this is fine, you already did help me a lot. If you do want to spend time on v2.5, that would be awesome! :-)

Yours,
John

**Comment by jobisoft** [comment]
I just played around with categories of exceptions. Testing against an O365 server. Even after fixing the pageswitch stuff, I still cannot clear categories in exceptions (with v2.5 I cannot even change categories).

I than tried to do the same thing on the O365 portal, but I cannot ??? The UI of https://outlook.office.com, does not let me change categories of a series if I edit a single element. How did you manage to do that?

There are two other EAS server implementation I test: horde and zpush (Kopano, Zarafa, ...)

If I test against a horde server, everything works as expected.
If I test against a zpush server, it does not throw an status 6, but seems to ignore any changes to categories.

Since I cannot identify the server type, I am thinking to drop category changes in exception by not sending them back to the server. Any objections?

**Comment by ChrisJAllan** [comment]
It seems like that's the best option. It would be great if Microsoft could follow their own documentation...

**Pull request merged**